### PR TITLE
Do not emit GEP instruction when pushing string literals to stack

### PR DESCRIPTION
**Pull request by michalgr**
I am able to crash bpftrace on one of asserts in llvm/Support/Casting.h when
building with LLVM8 and asserts enabled. Minimal repro I found is:
```
bpftrace -e 'BEGIN { printf("%s", "a"); }'
```
The crash occurs while creating GEP instruction when generating code pushing "a"
string literal to stack (see
https://github.com/iovisor/bpftrace/blob/master/src/ast/codegen_llvm.cpp#L46).
GEP expects its argument to be a `Value` of type `PointerType`, whereas type of
`Constant` pointed to by `const_str` is `ArrayType`. Proposed fix is to skip GEP
in this case, store opcode can handle arrays.

The issue might be hard to spot because with asserts disabled code seem to
work fine (fields of PointerType and ArrayType fortunately overlap in a way
that makes sense, `PointeeTy` in `PointerType` overlaps with `ContainedType`
added in `ArrayType`'s superclass).

Test plan:
Build bpftrace with LLVM8 and asserts enabled and make sure that I can't repro
the crash anymore. In additoin pick  run:
- biolatency.bt
- biosnoop.bt
- runqlat.bt
- tcpconnect.bt

### CODE REVIEW

**Comment by mmarchini** [comment]
Seems to make sense. Since the tests didn't fail I guess this won't change the generated IR in most cases, correct?

> Build bpftrace with LLVM8 and asserts enabled and make sure that I can't repro
the crash anymore. In additoin pick run

Is this a custom LLVM build with asserts enabled, or having asserts enabled on bpftrace is enough? I had problems in the past with LLVM debug build where we would hit some asserts as well.

**Comment by michalgr** [comment]
> Seems to make sense. Since the tests didn't fail I guess this won't change the generated IR in most cases, correct?

I think so. `bpftrace -e 'BEGIN { printf("%s", "abc"); }'` reports the same IR when I use preinstalled version (v0.9) and the one with this change and asserts enabled. After all getting address of the 0th element of an array is easy to optimize...

> Is this a custom LLVM build with asserts enabled, or having asserts enabled on bpftrace is enough? I had problems in the past with LLVM debug build where we would hit some asserts as well.

My setup is a bit inconsistent. I build everything on my own, llvm with NDEBUG defined, bcc and bpftrace without. Because of that casts in llvm's *.cpp files did not crash, but casts performed in functions defined in headers did. I guess some of the weak symbols were pulled from bcc/bpftrace. There definitely are casts in llvm's *.cpp files that should crash if llvm is built in debug mode, for example this one: https://github.com/llvm/llvm-project/blob/master/llvm/lib/IR/Constants.cpp#L1990-L1993

**Comment by michalgr** [comment]
I'll build LLVM in debug mode as well and give it a try. It would be a pity to trade one bad assert for another bad assert. Let's not merge until then ?

**Comment by mmarchini** [comment]
> It would be a pity to trade one bad assert for another bad assert. Let's not merge until then ?

I don't think this would introduce another bad assert, but I'm almost sure there are other bad asserts out there if you use Debug LLVM.

**Comment by michalgr** [comment]
This change makes the following assert fire: https://github.com/llvm/llvm-project/blob/release/8.x/llvm/lib/IR/Instructions.cpp#L1208. I'll do some casting before storing the value.

**Comment by michalgr** [comment]
Now `bpftrace -e 'BEGIN { printf("%s", "a"); }'` does not crash with all asserts enabled, but there are still many type bugs remaining.

File: src/ast/irbuilderbpf.cpp [hunk 1]
```diff
@@ -621,7 +621,13 @@ void IRBuilderBPF::CreatePerfEventOutput(Value *ctx, Value *data, size_t size)
   // int bpf_perf_event_output(ctx, map, flags, data, size)
   FunctionType *perfoutput_func_type = FunctionType::get(
       getInt64Ty(),
-      {getInt8PtrTy(), getInt8PtrTy(), getInt64Ty(), getInt8PtrTy(), getInt64Ty()},
+      {
+        getInt8PtrTy(),
+        getInt64Ty(),
```
Comment (ajor) [inline src/ast/irbuilderbpf.cpp]: `int bpf_perf_event_output(struct pt_reg *ctx, struct bpf_map *map, u64 flags, void *data, u64 size)`
This argument is meant to be a pointer

Comment (michalgr) [reply src/ast/irbuilderbpf.cpp]: Yep, but the argument we pass is always of type i64. Return type of llvm.bpf.pseudo is i64 and this is what we pass to bpf_perf_event_output. Should I add one more change where we'll define an opaque type "bpf_map" or something like that and cast results of llvm.bpf.pseudo to bpf_map* ?

**Comment by mmarchini** [comment]
> Now bpftrace -e 'BEGIN { printf("%s", "a"); }' does not crash with all asserts enabled, but there are still many type bugs remaining.

I tried this PR with a Debug build of LLVM 6 and got this error:

```
# ./src/bpftrace -e 'BEGIN { printf("%s", "a"); }'
Program aborted due to an unhandled Error:
Error value was Success. (Note: Success values must still be checked prior to being destroyed).
Aborted (core dumped)
```

I built clang/llvm with:

```
cmake ../llvm -DCMAKE_BUILD_TYPE=Debug -G Ninja -DLLVM_ENABLE_PROJECTS="clang;llvm" -DLLVM_TARGETS_TO_BUILD="X86;BPF" -DLLVM_BUILD_LLVM_DYLIB=ON -DLLVM_ENABLE_RTTI=ON
```

It still LGTM (maybe address https://github.com/iovisor/bpftrace/pull/667/files#r287598795), I don't think we need to fix all asserts in one PR.

**Comment by michalgr** [comment]
That argument should be a pointer now, but I am pretty sure I did not manage to fix all the tests (a lot of tests have up to 3 cases depending on LLVM version, I tested with llvm8). I'll rely on travis to catch tests I did not fix correctly.

Are there any tips on how to fix tests in cases like this? I modified all of them by hand incorporating feedback from bpftrace_test. That took a while.

**Comment by mmarchini** [comment]
Well, CI is happy. I'll run runtime tests and tooling tests later to see if there has been any changes there.

> Are there any tips on how to fix tests in cases like this?

Unfortunately no. Ideally we should have this automated somehow, but right now this has to be done manually.

@ajor is this PR good to merge for you?

**Comment by ajor** [comment]
I think it was fine last time I looked at the code, but I'd still like to go through the changed tests just to make sure nothing unexpected is happening there, so not ready just yet. Normally it's not too bad to have to manually update the codegen tests, since the changes should be checked manually anyway, but I agree it is a bit annoying when it's so many like this.

**Comment by mmarchini** [comment]
FWIW I had a vim bind a while back that would execute the test with `-d` and paste it in the codegen file, but that's not helpful when dealing with multiple LLVM versions or several test files.

**Comment by ajor** [comment]
Ah I forgot about pseudo just returning 64 bit ints. I actually prefer the code before the final commit since it's simpler to just use the ints directly - can we back that one out or do you feel strongly about it?

I'd be happy to fix the conflicts in this pull request since it's my fault this didn't get in before the version freeze. I'll do that this weekend if you don't have time before then.

**Comment by michalgr** [comment]
I won't be able to get back to this until later next week, if you feel like merging this sooner then please take over.

**Comment by michalgr** [comment]
I dropped the last commit and fixed conflicts.

**Comment by ajor** [comment]
Took a while to merge, but on the plus side you got the 1000th commit according to Github ;)

**Comment by michalgr** [comment]
Thank you !

> you got the 1000th commit according to Github ;)

🍾 I am considering printing this on a tshirt :)

**Pull request merged**